app/write_square_test.go

```diff
@@ -0,0 +1,145 @@
+package app
+
+import (
+	"bytes"
+	"sort"
+	"testing"
+
+	"github.com/celestiaorg/celestia-app/pkg/shares"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	tmlog "github.com/tendermint/tendermint/libs/log"
+	core "github.com/tendermint/tendermint/proto/tendermint/types"
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+func Test_finalizeLayout(t *testing.T) {
+	ns1 := []byte{1, 1, 1, 1, 1, 1, 1, 1}
+	ns2 := []byte{2, 2, 2, 2, 2, 2, 2, 2}
+	ns3 := []byte{3, 3, 3, 3, 3, 3, 3, 3}
+
+	type test struct {
+		squareSize      uint64
+		nonreserveStart int
+		ptxs            []parsedTx
+		expectedIndexes []uint32
+	}
+	tests := []test{
+		{
+			squareSize:      4,
+			nonreserveStart: 10,
+			ptxs: generateParsedTxsWithNIDs(
+				[][]byte{ns1},
+				[]int{1},
+			),
+			expectedIndexes: []uint32{10},
+		},
+		{
+			squareSize:      4,
+			nonreserveStart: 10,
+			ptxs: generateParsedTxsWithNIDs(
+				[][]byte{ns1, ns1},
+				[]int{100, 100},
+			),
+			expectedIndexes: []uint32{10, 11},
+		},
+		{
+			squareSize:      4,
+			nonreserveStart: 10,
+			ptxs: generateParsedTxsWithNIDs(
+				[][]byte{ns1, ns1, ns1, ns1, ns1, ns1, ns1, ns1, ns1, ns1},
+				[]int{100, 100, 100, 100, 100, 100, 100, 100, 100, 100},
+			),
+			expectedIndexes: []uint32{10, 11, 12, 13, 14, 15},
+		},
+		{
+			squareSize:      4,
+			nonreserveStart: 6,
+			ptxs: generateParsedTxsWithNIDs(
+				[][]byte{ns1, ns1, ns1, ns1, ns1, ns1, ns1, ns1, ns1, ns1},
+				[]int{100, 100, 100, 100, 100, 100, 100, 100, 100, 100},
+			),
+			expectedIndexes: []uint32{6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+		},
+		{
+			squareSize:      4,
+			nonreserveStart: 3,
+			ptxs: generateParsedTxsWithNIDs(
+				[][]byte{ns1, ns1, ns1},
+				[]int{10000, 10000, 1000000},
+			),
+			expectedIndexes: []uint32{},
```

Comment:
[question] is this test verifying that a square size of 4 can't contain a 10,000 byte blob because 10,000 / 503 (sparse share content size) is 19.88071571 and only 16 shares exist in a square size of 4?

Reply:
yeah exactly